packages/contracts/contracts/discovery/ProjectVoting.sol

```diff
@@ -0,0 +1,245 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.12;
+
+import {ICommon} from "./ICommon.sol";
+import {Math} from "../libraries/Math.sol";
+
+import "hardhat/console.sol";
+
+abstract contract ProjectVoting is ICommon {
+    /// number of votes for each project
+    uint32[] public votes;
+
+    /// Ordered list of slot winners
+    /// @dev only valid up to the last time there was a vote. Use getWinners() instead.
+    address[] winners;
+
+    /// index in the votes array to project address
+    mapping(uint256 => address) public votesIndexToProject;
+
+    /// project address to index  in the votes array
+    mapping(address => uint256) public projectToVotesIndex;
+
+    /// user => votes
+    mapping(address => uint256) public userVoteCount;
+
+    /// project address => votes
+    mapping(address => uint256) public projectVoteCount;
+
+    /// project address => votes
+    mapping(address => uint256) public weightedProjectVoteCount;
+
+    /// project => (user => voted)
+    mapping(address => mapping(address => bool)) public userVotesPerProject;
+
+    /// Project address to status
+    /// @dev only valid up to the last time there was a vote. Use getProject(address) instead.
+    mapping(address => ProjectStatus) projectStatuses;
+
+    function projectVoting_projects()
+        public
+        view
+        virtual
+        returns (address[] memory);
+
+    function projectVoting_votingPeriod()
+        public
+        view
+        virtual
+        returns (Period memory);
+
+    function projectVoting_singleSlotDuration()
+        public
+        view
+        virtual
+        returns (uint256);
+
+    function projectVoting_initialBonus() public view virtual returns (int256);
+
+    function projectVoting_finalBonus() public view virtual returns (int256);
+
+    enum ProjectStatus {
+        InProgress,
+        Won,
+        Lost
+    }
+
+    struct Vote {
+        uint256 originalIndex;
+        uint32 count;
+    }
+
+    function _vote(address projectAddress) internal {
+        require(
+            !userVotesPerProject[projectAddress][msg.sender],
+            "already voted in this project"
+        );
+        _defineWinners();
+        require(
+            projectStatuses[projectAddress] == ProjectStatus.InProgress,
+            "project is not in progress"
+        );
+        userVotesPerProject[projectAddress][msg.sender] = true;
+        userVoteCount[msg.sender]++;
+        projectVoteCount[projectAddress]++;
+        weightedProjectVoteCount[projectAddress] += _calculateWeightedVote(
+            block.timestamp
+        );
+
+        if (
+            votes.length == 0 ||
+            (projectToVotesIndex[projectAddress] == 0 &&
+                votesIndexToProject[0] != projectAddress)
+        ) {
```

Comment:
what if, instead of this logic, we just receive `address[] projects` in the constructor here, and populate `votesIndexToProject` and `projectToVotesIndex` there?

Reply:
How do you mean? Are you suggesting we populate the `votes` array with a `0` for each project? Because right now the length of the array is equal to the amount of projects that received votes, not the total number of projects.

Reply:
yes, that's what I'm suggesting. to avoid having this logic here on each individual vote.